### DISCUSSION TITLE
Call to `metadata.create_all` with `bind=session.bind` causes session not to commit - bug or misuse?

### DISCUSSION THREAD

**Discussion by nick8325** [sqlalchemy/sqlalchemy#10935] [Usage Questions] [answered]
I've bumped into some strange behaviour on SQLAlchemy 2.0.25 and am wondering whether this is a bug, or whether I'm misusing the API. In short, a transaction is being rolled back when I would expect it to be committed.

Here is some test code which just creates a table and does an insertion, using a `Session`:

```python
from sqlalchemy import Table, Column, Integer, MetaData, insert, create_engine
from sqlalchemy.orm import Session

metadata = MetaData()
test = Table("test", metadata, Column("id", Integer))

engine = create_engine("sqlite+pysqlite:///db", echo=True, future=True)
        
with engine.connect() as conn:
    with Session(bind=conn) as session:
        metadata.create_all(bind=session.bind)
        session.execute(insert(test).values(id=1))
        session.commit()
```

I run the code once to create the table, then I run it again. I get the following log output:

```
2024-01-29 14:36:10,612 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-01-29 14:36:10,612 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("test")
2024-01-29 14:36:10,612 INFO sqlalchemy.engine.Engine [raw sql] ()
2024-01-29 14:36:10,613 INFO sqlalchemy.engine.Engine INSERT INTO test (id) VALUES (?)
2024-01-29 14:36:10,613 INFO sqlalchemy.engine.Engine [generated in 0.00009s] (1,)
2024-01-29 14:36:10,614 INFO sqlalchemy.engine.Engine ROLLBACK
```

You can see that the transaction was rolled back even though I called `session.commit()`. Note, I get the same result if I change `bind=session.bind` to `bind=conn`.

If I comment out the line `metadata.create_all(bind=session.bind)`, or change `bind=session.bind` to `bind=engine`, then the transaction is committed as expected:

```
2024-01-29 14:42:59,570 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-01-29 14:42:59,571 INFO sqlalchemy.engine.Engine INSERT INTO test (id) VALUES (?)
2024-01-29 14:42:59,571 INFO sqlalchemy.engine.Engine [generated in 0.00012s] (1,)
2024-01-29 14:42:59,571 INFO sqlalchemy.engine.Engine COMMIT
```

**Question:** is this a bug, or is it expected that I need to use `bind=engine` rather than `bind=session.bind`/`bind=conn` when I call `metadata.create_all`?

----

P.S. I looked into the code a bit to figure out why `metadata.create_all` is "preventing" the later `session.commit` from happening. When `session.execute` is executed, `Session._connection_for_bind` is called. However, when there was a previous call to `metadata.create_all(bind=session.bind)`, then `conn.in_transaction()` is true, which results in `_connection_for_bind` setting the flag `should_commit` to `False` [here](https://github.com/sqlalchemy/sqlalchemy/blob/582058ea54ba9445c7bb0ffd94690f77ce55a5ba/lib/sqlalchemy/orm/session.py#L1184).

**Comment by zzzeek** [comment, marked as answer]
well I wouldn't use the connection in this way, just to eliminate confusion, but yes, everything you see is "expected".

example one: connection is used before session gets to do anything:

```py
with engine.connect() as conn:
    with Session(bind=conn) as session:
        metadata.create_all(bind=conn)
        session.execute(insert(test).values(id=1))
        session.commit()
```

engine.connect() returns a connection.  session(bind=conn) does nothing with the connection, no transaction is started.   create_all() runs on connection, this starts a transaction.  connection is now in a transaction.   session.execute() is called, session looks at this connection for the first time, sees that it's already in a transaction, and goes into [rollback_only](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.params.join_transaction_mode) mode (which is now configurable in 2.0, however see the docs there referring to what 1.4 would do in all cases) where it assumes you're in a test suite with an external transaction and that commit() should not actually commit this transaction.   hence nothing happens.   engine.connect() block ends, rollback is emitted.  no bug

example two: session knows about the connection **before** anything else is done with it:

```py
with engine.connect() as conn:
    with Session(bind=conn) as session:
        session.connection()
        metadata.create_all(bind=conn)
        session.execute(insert(test).values(id=1))
        session.commit()
```

above, the Session when asked for a connection starts its transaction on a clean connection that was not already in a transaction, in the equivalent of [control_fully](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.params.join_transaction_mode) mode.    create_all() is then called on that connection, runs the SQL, does not affect the Session:

```
CREATE TABLE test (
	id INTEGER
)


2024-01-29 09:28:02,781 INFO sqlalchemy.engine.Engine [no key 0.00004s] ()
2024-01-29 09:28:02,782 INFO sqlalchemy.engine.Engine INSERT INTO test (id) VALUES (?)
2024-01-29 09:28:02,782 INFO sqlalchemy.engine.Engine [generated in 0.00008s] (1,)
2024-01-29 09:28:02,782 INFO sqlalchemy.engine.Engine COMMIT

```

What would the *correct* way to do this, given you want to run `create_all()` within the same transaction that the session is using.   Always access this connection using [Session.connection()](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.connection) so that the Session is included in the usage sequence for the connection:

```py
with engine.connect() as conn:
    with Session(bind=conn) as session:
        metadata.create_all(bind=session.connection())
        session.execute(insert(test).values(id=1))
        session.commit()
```

**Reply by nick8325** [reply]
Thanks a lot for the clear explanation! That makes a lot of sense - so when I create a session on a connection that was already used, it goes into `rollback_only` mode on the assumption that the transaction is being managed outside of the session. When I switch to using `session.connection()` then everything works.

**Reply by CaselIT** [reply]
Note that you can control how the connection behaves using the https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.params.join_transaction_mode option of a session

**Reply by nick8325** [reply]
Thanks!
